### PR TITLE
Add old intel and amd cpus

### PR DESCRIPTION
**Pull request by timw4mail**
All of these values come from sandpile.org. 

- Add Intel 486
- Add Intel Pentium Pro
- Add AMD 486
- Add AMD K5

Questions:
1. Do you have a different way to better structure the list of inferred CPU names
2. Do we need a backup way to determine cpu speed?
3. Should we hide SSE, Peak Performance, etc. line items if unknown?

### CODE REVIEW

File: src/x86/uarch.c [hunk 1]
```diff
@@ -149,16 +153,30 @@ struct uarch* get_uarch_from_cpuid_intel(uint32_t ef, uint32_t f, uint32_t em, u
...
   CHECK_UARCH(arch, 0,  5,  0,  0, NA, "P5",                UARCH_P5,              800)
   CHECK_UARCH(arch, 0,  5,  0,  1, NA, "P5",                UARCH_P5,              800)
-  CHECK_UARCH(arch, 0,  5,  0,  2, NA, "P5",                UARCH_P5,              UNK)
-  CHECK_UARCH(arch, 0,  5,  0,  3, NA, "P5",                UARCH_P5,              600)
-  CHECK_UARCH(arch, 0,  5,  0,  4, NA, "P5 (MMX)",          UARCH_P5_MMX,          UNK)
-  CHECK_UARCH(arch, 0,  5,  0,  7, NA, "P5 (MMX)",          UARCH_P5_MMX,          UNK)
-  CHECK_UARCH(arch, 0,  5,  0,  8, NA, "P5 (MMX)",          UARCH_P5_MMX,          250)
+  CHECK_UARCH(arch, 0,  5,  0,  2, NA, "P54C",              UARCH_P5,              UNK)
+  CHECK_UARCH(arch, 0,  5,  0,  3, NA, "P24T (Overdrive)",  UARCH_P5,              600) // sandpile.org
+  CHECK_UARCH(arch, 0,  5,  0,  4, NA, "P55C (MMX)",        UARCH_P5_MMX,          350)
```
Comment (Dr-Noob) [inline src/x86/uarch.c:169]: If you have the source for the 350 please add it as a comment

File: src/x86/uarch.c [hunk 2]
```diff
@@ -149,16 +153,30 @@ struct uarch* get_uarch_from_cpuid_intel(uint32_t ef, uint32_t f, uint32_t em, u
...
+  CHECK_UARCH(arch, 0,  5,  0,  7, NA, "P55C (MMX)",        UARCH_P5_MMX,          350)
+  CHECK_UARCH(arch, 0,  5,  0,  8, NA, "P55C (MMX)",        UARCH_P5_MMX,          250)
   CHECK_UARCH(arch, 0,  5,  0,  9,  0, "Lakemont",          UARCH_LAKEMONT,         32)
   CHECK_UARCH(arch, 0,  5,  0,  9, NA, "P5 (MMX)",          UARCH_P5_MMX,          UNK)
   CHECK_UARCH(arch, 0,  5,  0, 10,  0, "Lakemont",          UARCH_LAKEMONT,         32)
+  CHECK_UARCH(arch, 0,  6,  0,  1,  1, "P6",                UARCH_P6_PRO,          UNK)
```
Comment (Dr-Noob) [inline src/x86/uarch.c:175]: Shouldn't this be `"P6 Pro"`?

File: src/x86/uarch.c [hunk 3]
```diff
@@ -149,16 +153,30 @@ struct uarch* get_uarch_from_cpuid_intel(uint32_t ef, uint32_t f, uint32_t em, u
...
+  CHECK_UARCH(arch, 0,  6,  0,  1,  2, "P6",                UARCH_P6_PRO,          600)
```
Comment (Dr-Noob) [inline src/x86/uarch.c:176]: Same, if you have a source for the 600 please add it.

File: src/x86/uarch.c [hunk 4]
```diff
@@ -149,16 +153,30 @@ struct uarch* get_uarch_from_cpuid_intel(uint32_t ef, uint32_t f, uint32_t em, u
   // ------------------------------------------------------------------------------- //
   //                EF  F  EM   M   S                                                //
   UARCH_START
+  CHECK_UARCH(arch, 0,  4,  0,  0, NA, "i80486DX",          UARCH_I486,            UNK) // sandpile.org
```
Comment (Dr-Noob) [inline src/x86/uarch.c:156]: Can you provide the exact sandpile.org URL that has these CPU values?

**Comment by Dr-Noob** [comment]
Thanks for the patch. Overall, looks good. I've added a few comments. Also, could you provide some screenshot to show how this looks on such old Intel/AMD CPUs? This will also be useful to better answer your questions:

> 1. Do you have a different way to better structure the list of inferred CPU names

Not right now, but because it's size is increasing I will probably rewrite it into a more structured way (similar to `CHECK_UARCH` but for CPU names). Feel free to do this if you want.

> 2. Do we need a backup way to determine cpu speed?

It would be good to have one if needed, but I would need to see some evidences. Do you have any CPU for which you cannot find the CPU frequency?

> 3. Should we hide SSE, Peak Performance, etc. line items if unknown?

Peak Performance no, for SSE maybe. Do you have any CPU in such scenario?

**Review by Dr-Noob — CHANGES_REQUESTED** [review]

File: src/x86/uarch.c (hunk 4 above)
Comment (timw4mail) [reply src/x86/uarch.c:156]: https://sandpile.org/x86/cpuid.htm#level_0000_0001h

Under the 'model' section of the giant table

File: src/x86/uarch.c (hunk 2 above)
Comment (timw4mail) [reply src/x86/uarch.c:175]: This is to separate the Pentium Pro (the original P6) from Pentium 2/Pentium 3. The inferred name lookup will show "Pentium Pro".

File: src/x86/uarch.c (hunk 3 above)
Comment (timw4mail) [reply src/x86/uarch.c:176]: I don't seem to have a proper source. I'll go through the Pentium Pros and find a source or update to unknown.

**Comment by timw4mail** [comment]
Before the last set of merges to master, I saw missing clock speeds a lot more often. Now I see empty SSE:

PXL_20240802_012957338

File: src/x86/uarch.c (hunk 4 above)
Comment (Dr-Noob) [reply src/x86/uarch.c:156]: Right, but can you add it as a comment? It's important to keep these things documented.

**Comment by Dr-Noob** [comment]
> Before the last set of merges to master, I saw missing clock speeds a lot more often. Now I see empty SSE:
> 
> PXL_20240802_012957338

Regarding the frequency, I have implemented a new functionality recently that estimates the maximum frequency. That's why you have noticed this change in the recent commits.

Regarding SSE, I'm going to push a fix, just rebase and confirm that it works as expected.

**Comment by Dr-Noob** [comment]
There is something important I forgot to mention in the last review. 

Setting different strings for the same uarch, like in:

```c
  CHECK_UARCH(arch, 0,  4,  0,  1, NA, "i80486DX-50",       UARCH_I486,            UNK)
  CHECK_UARCH(arch, 0,  4,  0,  2, NA, "i80486SX",          UARCH_I486,            UNK)
  CHECK_UARCH(arch, 0,  4,  0,  3, NA, "i80486DX2",         UARCH_I486,            UNK)
  CHECK_UARCH(arch, 0,  4,  0,  4, NA, "i80486SL",          UARCH_I486,            UNK)
  CHECK_UARCH(arch, 0,  4,  0,  5, NA, "i80486SX2",         UARCH_I486,            UNK)
  CHECK_UARCH(arch, 0,  4,  0,  7, NA, "i80486DX2WB",       UARCH_I486,            UNK)
  CHECK_UARCH(arch, 0,  4,  0,  8, NA, "i80486DX4",         UARCH_I486,            UNK)
  CHECK_UARCH(arch, 0,  4,  0,  9, NA, "i80486DX4WB",       UARCH_I486,            UNK)
```

This is incorrect! All strings for a given uarch must be the same (in this case it should be something like `i486` for all `UARCH_I486`, as this is the name of the microarchitecture). I understand that you want to differentiate between CPUs here, but this would be the CPU name, not the microarchitecture. The right thing to do is:

1. Save the CPUID value in `struct cpuInfo`.
2. Use these values in `infer_cpu_name_from_uarch` to determine the exact CPU name depending on those values (rather than incorrectly doing so in `get_uarch_from_cpuid_intel`).

You can implement this yourself, or just leave it like this and I'll implement it after merging your PR.

**Comment by Dr-Noob** [comment]
Another (optional) enhancement to your PR is to add support for detecting [MMX](https://es.wikipedia.org/wiki/MMX).

As you showed, your CPU does not support SSE but I'll probably support MMX. You could take the AVX or SSE detection as a reference to implement this. After that, in your case it would show that MMX is supported for your CPU (assuming it does support it) and also the peak performance would be higher.

**Comment by timw4mail** [comment]
> There is something important I forgot to mention in the last review.
> 
> Setting different strings for the same uarch, like in:
> 
> ```c
>   CHECK_UARCH(arch, 0,  4,  0,  1, NA, "i80486DX-50",       UARCH_I486,            UNK)
>   CHECK_UARCH(arch, 0,  4,  0,  2, NA, "i80486SX",          UARCH_I486,            UNK)
>   CHECK_UARCH(arch, 0,  4,  0,  3, NA, "i80486DX2",         UARCH_I486,            UNK)
>   CHECK_UARCH(arch, 0,  4,  0,  4, NA, "i80486SL",          UARCH_I486,            UNK)
>   CHECK_UARCH(arch, 0,  4,  0,  5, NA, "i80486SX2",         UARCH_I486,            UNK)
>   CHECK_UARCH(arch, 0,  4,  0,  7, NA, "i80486DX2WB",       UARCH_I486,            UNK)
>   CHECK_UARCH(arch, 0,  4,  0,  8, NA, "i80486DX4",         UARCH_I486,            UNK)
>   CHECK_UARCH(arch, 0,  4,  0,  9, NA, "i80486DX4WB",       UARCH_I486,            UNK)
> ```
> 
> This is incorrect! All strings for a given uarch must be the same (in this case it should be something like `i486` for all `UARCH_I486`, as this is the name of the microarchitecture). I understand that you want to differentiate between CPUs here, but this would be the CPU name, not the microarchitecture. The right thing to do is:
> 
>     1. Save the CPUID value in `struct cpuInfo`.
> 
>     2. Use these values in `infer_cpu_name_from_uarch` to determine the exact CPU name depending on those values (rather than incorrectly doing so in `get_uarch_from_cpuid_intel`).
> 
> 
> You can implement this yourself, or just leave it like this and I'll implement it after merging your PR.

That does make sense. 

So, there are three different things conflated in the current mapping

1. Microarchitecture name
2. More specific model name (486DX, 486SX, etc)
3. Core codename when different from the microarchitecture. name (Klamath, Deschutes, etc)

To keep the implementation purely for uarch, this is my thought:

1. Add a new function that takes cpuid values and uarch to potentially get a more detailed model name. 
    For the current line `CHECK_UARCH(arch, 0,  4,  0,  0, NA, "i80486DX",          UARCH_I486,            UNK)`, this new function would return 'DX', so the full string might be "Intel 486 DX".
2. Add an attribute for the CPU core codename, maybe "Codename" or "Core name". 

> Another (optional) enhancement to your PR is to add support for detecting [MMX](https://es.wikipedia.org/wiki/MMX).
> 
> As you showed, your CPU does not support SSE but I'll probably support MMX. You could take the AVX or SSE detection as a reference to implement this. After that, in your case it would show that MMX is supported for your CPU (assuming it does support it) and also the peak performance would be higher.

Adding MMX detection does seem like a good idea. (Although the Pentium Pro does predate MMX). I think that should wait for a different PR.

File: src/x86/uarch.c (hunk 4 above)
Comment (timw4mail) [reply src/x86/uarch.c:156]: Sure. Do you just want each of those CPUs to have a comment with the same URL?

**Comment by timw4mail** [comment]
Also, I did see that the SSE fix worked.

**Comment by Dr-Noob** [comment]
> > There is something important I forgot to mention in the last review.
> > Setting different strings for the same uarch, like in:
> > ```c
> >   CHECK_UARCH(arch, 0,  4,  0,  1, NA, "i80486DX-50",       UARCH_I486,            UNK)
> >   CHECK_UARCH(arch, 0,  4,  0,  2, NA, "i80486SX",          UARCH_I486,            UNK)
> >   CHECK_UARCH(arch, 0,  4,  0,  3, NA, "i80486DX2",         UARCH_I486,            UNK)
> >   CHECK_UARCH(arch, 0,  4,  0,  4, NA, "i80486SL",          UARCH_I486,            UNK)
> >   CHECK_UARCH(arch, 0,  4,  0,  5, NA, "i80486SX2",         UARCH_I486,            UNK)
> >   CHECK_UARCH(arch, 0,  4,  0,  7, NA, "i80486DX2WB",       UARCH_I486,            UNK)
> >   CHECK_UARCH(arch, 0,  4,  0,  8, NA, "i80486DX4",         UARCH_I486,            UNK)
> >   CHECK_UARCH(arch, 0,  4,  0,  9, NA, "i80486DX4WB",       UARCH_I486,            UNK)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This is incorrect! All strings for a given uarch must be the same (in this case it should be something like `i486` for all `UARCH_I486`, as this is the name of the microarchitecture). I understand that you want to differentiate between CPUs here, but this would be the CPU name, not the microarchitecture. The right thing to do is:
> > ```
> > 1. Save the CPUID value in `struct cpuInfo`.
> > 
> > 2. Use these values in `infer_cpu_name_from_uarch` to determine the exact CPU name depending on those values (rather than incorrectly doing so in `get_uarch_from_cpuid_intel`).
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You can implement this yourself, or just leave it like this and I'll implement it after merging your PR.
> 
> That does make sense.
> 
> So, there are three different things conflated in the current mapping
> 
> 1. Microarchitecture name
> 2. More specific model name (486DX, 486SX, etc)
> 3. Core codename when different from the microarchitecture. name (Klamath, Deschutes, etc)
> 
> To keep the implementation purely for uarch, this is my thought:
> 
> 1. Add a new function that takes cpuid values and uarch to potentially get a more detailed model name.
>    For the current line `CHECK_UARCH(arch, 0,  4,  0,  0, NA, "i80486DX",          UARCH_I486,            UNK)`, this new function would return 'DX', so the full string might be "Intel 486 DX".

If prefer not to use the uarch name directly for the CPU name. What I mean is that I would rather return "Intel 486 DX" instead of "DX" and then append that to the "Intel 486". Anyway, I will implement this after your PR lands.

> 2. Add an attribute for the CPU core codename, maybe "Codename" or "Core name".

This is actually a very good point. I want to implement this since 2021! (see #120) but not much time to spend on this.

> > Another (optional) enhancement to your PR is to add support for detecting [MMX](https://es.wikipedia.org/wiki/MMX).
> > As you showed, your CPU does not support SSE but I'll probably support MMX. You could take the AVX or SSE detection as a reference to implement this. After that, in your case it would show that MMX is supported for your CPU (assuming it does support it) and also the peak performance would be higher.
> 
> Adding MMX detection does seem like a good idea. (Although the Pentium Pro does predate MMX). I think that should wait for a different PR.

Agreed, can be good for a different PR if you have a good testcase (a CPU that supports MMX but not newer instructions like SSE). Right now this PR looks good to me. If you can confirm that this works for your hardware I'll merge into master soon :+1:

File: src/x86/uarch.c (hunk 4 above)
Comment (Dr-Noob) [reply src/x86/uarch.c]: Having each one with a comment pointing to the same URL (what you have right now) is fine.

**Review by Dr-Noob — APPROVED** [review]

**Comment by timw4mail** [comment]
For the CPU identification, everything looks good to me. 

The only minor issue (which we can figure out later) is that getting a clock speed for these old CPUs is tricky: 

- For 486 chips, there's no timestamp counter, nor a speed reference from the linux kernel. 
- It looks like `/proc/cpuinfo` could be a good fallback. It might even be more accurate than the current fallback, which always seems about 20MHz slow on the Pentium Pro system

### 486 Socket
PXL_20240805_201127593
PXL_20240805_224642871
PXL_20240805_191739398
PXL_20240805_195244417 MP
### Pentium Pro
PXL_20240806_151553343
### AMD K5
PXL_20240806_154918404
PXL_20240806_155054386
PXL_20240806_160040916 MP

**Comment by Dr-Noob** [comment]
Okay, I'm ready to merge. Just wanted to confirm if you are happy with this?

Screenshot_20240818_163818

**Comment by Dr-Noob** [comment]
Sorry for the big delay. Sometimes it's dificult to find time for this. About the frequency, I don't think there is much to do. I'm afraid that without `perf_event_open` the accuracy would be poor. We could rely on `/proc/cpuinfo` however, to read the frequency, that's a good point. I'll schedule some time to do some test to see if this method is reliable :+1:

**Comment by timw4mail** [comment]
Yeah, I'm good to merge.

**Comment by Dr-Noob** [comment]
Merged! Thank you for your contribution and looking forward for the next :+1:

**Pull request merged**